### PR TITLE
Add improved tab completion functionality

### PR DESCRIPTION
**Pull request by Rudedog9d**
Add improved tab completion function to `completers.py`

### CODE REVIEW

File: pypsi/completers.py [hunk 1]
```diff
@@ -22,6 +22,78 @@
 from pypsi.os import path_completer  # noqa
 
 
+def command_completer(command, shell, args, prefix, case_sensitive=False):
+    '''
+    Completion function that can tab complete options,
+    options' values, and positional paramaters.
+    :param command: The action object returned from
```
Comment (ameily) [inline pypsi/completers.py]: Include the argument type in the docstring. ie:

```python
'''
...
:param pypsi.shell.Shell shell: the current Shell object
:param str prefix: The partial arg being completed
...
'''
'''

File: pypsi/completers.py [hunk 2]
```diff
@@ -22,6 +22,78 @@
...
+            base = [key for key in command.choices]
+            completions.extend([x for x in base
+                                if x.startswith(prefix) or not prefix])
+            return sorted(completions)
+        if command.choices:
+            # Get the parser for the current subcmd
+            cmd_parser = command.choices.get(args[0], None)
+            offset = 1  # Set an offset so the subcmd is not counted in index
+            last_arg = args[-2]
+    except AttributeError:
```
Comment (ameily) [inline pypsi/completers.py]: What can raise an `AttributeError`?

File: pypsi/core.py [hunk 3]
```diff
@@ -250,6 +250,12 @@ class PypsiArgParser(argparse.ArgumentParser):
       error stream
     '''
 
+    def __init__(self, *args, **kwargs):
+        self._pos_callbacks = []
```
Comment (ameily) [inline pypsi/core.py]: Add docstrings for these attributes:

```python
#: Positional something blah something
self._pos_callbacks = []
```

**Comment by ameily** [comment]
@Rudedog9d this is a pretty significant change so I'd like you to add more documentation for both the `PypsiArgParser` class and the `command_completer` function. Include how to properly take advantage of the option tab completion for both situations of a subparser and not a subparser.

File: pypsi/completers.py (hunk 2 above)
Comment (Rudedog9d) [inline pypsi/completers.py]: I believe `command.choices` will if command is a parser and not a subparser

**Comment by Rudedog9d** [comment]
@ameily I will go through and document everything better. Do you want me to add a tab-completion example to README.md? And/or maybe an example command using it?

**Comment by ameily** [comment]
> Do you want me to add a tab-completion example to README.md? And/or maybe an example command using it?

That would be great!

File: pypsi/core.py [hunk 4]
```diff
@@ -266,6 +272,89 @@ def print_help(self, file=None):
...
+        '''
+        Check if the optional argument has a value associated with it.
+        :param arg: Optional argument to check
+        :return: True if arg has a value, false otherwise
+        '''
+        val_actions = [argparse._AppendAction, argparse._StoreAction]
+        # Try catch in case arg doesn't exist
+        try:
+            # If a option's action is 'store' or 'append' is stores a value
+            return type(self._option_string_actions[arg]) in val_actions
```
Comment (ameily) [inline pypsi/core.py]: I'm not sure what `_option_string_actions` or  `_StoreAction/_AppendAction` are. Can you add a comment here saying the type you expect it to be and what its significance is?

File: pypsi/core.py [hunk 5]
```diff
@@ -266,6 +272,89 @@ def print_help(self, file=None):
...
+        except Exception as e:
+            print(e)
+            return False
+
+    def get_positional_callback(self, pos):
+        '''
```
Comment (ameily) [inline pypsi/core.py]: Almost everything in programming is assumed to be index-0. Does it make sense to change this function to be index-0 as well?

File: pypsi/core.py [hunk 6]
```diff
@@ -266,6 +272,89 @@ def print_help(self, file=None):
...
+        try:
+            # To get FIRST callback, return 0th list item
+            return self._pos_callbacks[pos - 1]
+        except IndexError:
+            if self._repeating_cb:
+                return self._repeating_cb
+            return None
+
+    def get_index(self, args):
```
Comment (ameily) [inline pypsi/core.py]: `get_index` isn't very descriptive. Maybe something like `get_positional_arg_index`?

File: pypsi/core.py [hunk 7]
```diff
@@ -266,6 +272,89 @@ def print_help(self, file=None):
         print(AnsiCodes.yellow, self.format_help(), AnsiCodes.reset, sep='',
               file=f)
 
+    def get_options(self):
```
Comment (ameily) [inline pypsi/core.py]: What is an `option` in the context of `get_options` and `get_option_callback`?

Comment (Rudedog9d) [inline pypsi/core.py]: Optional arguments (things that begin with a '-')

File: pypsi/core.py (hunk 5 above)
Comment (Rudedog9d) [inline pypsi/core.py]: I was back and forth on this. Coming back to it, I agree, and have changed it to index 0.

File: pypsi/completers.py [hunk 8]
```diff
@@ -22,6 +22,82 @@
 from pypsi.os import path_completer  # noqa
 
 
+def command_completer(parser, shell, args, prefix, case_sensitive=False):
+    '''
+    Completion function that can tab complete options,
```
Comment (ameily) [inline pypsi/completers.py]: Mention here that the developer should call this function with the same arguments passed into the command's `complete()` method.

File: pypsi/core.py [hunk 9]
```diff
@@ -266,6 +280,101 @@ def print_help(self, file=None):
...
+                    # reduce index to not count it's value as a pos param
+                    index -= 1
+            else:
+                # Is a positional param or value for an optional argument
+                index += 1
+
+        # return zero-based index
+        return index - 1
+
+    def add_argument(self, *args, **kwargs):
```
Comment (ameily) [inline pypsi/core.py]: To aid IDEs you can add your new arguments prior to the **kwargs:

```python
def add_argument(self, *args, callback=None, **kwargs):
    pass
```

File: pypsi/core.py [hunk 10]
```diff
@@ -266,6 +280,101 @@ def print_help(self, file=None):
...
+        :return: True if arg has a value, false otherwise
+        '''
+        val_actions = [argparse._AppendAction, argparse._StoreAction]
+        # _option_string_actions is a dictionary containing all of the optional
+        # arguments and the argparse action they should perform. Currently, the
+        # only two actions that store a value are _AppendAction/_StoreAction.
+        # These represent the value passed to 'action' in add_argument:
+        # parser.add_argument('-l', '--long', action='store')
+        action = self._option_string_actions.get(arg, None)
+        return type(action) in val_actions
```
Comment (ameily) [inline pypsi/core.py]: A more Pythonic way would be to use

```python
return isinstance(action, (argparse._AppendAction, argparse._StoreAction))
```

**Comment by Rudedog9d** [comment]
@ameily Do you have a unit test you suggest I look at as an example to create a test for this?

File: README.rst [hunk 11]
```diff
@@ -211,6 +211,53 @@ new commands in the shell.
...
+the possible completions can be specified. The callback function will be called
+from ``command_completer`` anytime tab is pressed while the user is currently
+entering that argument's value. Ex:
+
+.. code-block:: python
+
+    # Snippet from macro.py
+    self.parser.add_argument(
+         '-s', '--show', help='print macro body',
+         metavar='NAME', callback=self.complete_macros
```
Comment (ameily) [inline README.rst]: I think this needs to be `completer=....`.

**Pull request merged**